### ISSUE TITLE
задача про матрицы

### ISSUE THREAD

**Issue by uliss4ever** [uliss4ever/My_repository_3#2]
@Lex-2008
Вот она :)
https://leetcode.com/problems/count-square-submatrices-with-all-ones/
Соображения у меня пока такие:
Идем по массиву и проверяем все подмассивы в форме квадрата. И в каждом находим минимальное число. Если оно 0 - тогда идём дальше. Если 1 - значит в этом квадрате у нас все единицы, и счетчик увеличиваем на 1. Дальше надо увеличить высоту и ширину на 1 и так далее, пока в границу основного массива не упремся. 
"Более лучшая" мысль такая - что не надо проверять все квадраты, если он начинается (?) с нуля, то уже точно не подходит, идём дальше. Но у меня просто не получается добраться до всех квадратов. Вот тут на скрине просто посчитала сумму всего в каждом из подмассивов размером 2*2. Сейчас мне кажется, что минимум лучше искать, чем сумму, но по большому счету это не так важное, наверное. Главное пройти их все. Ну или пройти весь массив, ненужные при этом пропустить. А пока я разобралась только с теми, которые размером 2*2. 

![Снимок экрана 2021-03-11 в 15 02 01](https://user-images.githubusercontent.com/32711430/110784464-c785fc80-827a-11eb-80b8-648c0cce4e5c.png)

**Comment by Lex-2008** [comment]
Да, всё правильно. То, что у тебя написано на 12 строчке, я бы выделил в отдельную функцию: "посчитать сумму всего в подмассиве размером 2\*2, который начинается в элементе (i,j).". Теперь напиши такую же функцию для подмассивов размером 3\*3 и 4\*4 - и тогда, может быть, сможешь написать такую же функцию для квадратного подмассива произвольного размера n\*n?

**Comment by uliss4ever** [comment]
Функция вышла какая-то кривая. Я вышла за границы (out of range), а поскольку циклы в цикле, то запуталась...
В итоге всё стерла и подумала немного по-другому, как считаешь, в этом направлении может получиться?

![Снимок экрана 2021-03-12 в 11 48 01](https://user-images.githubusercontent.com/32711430/110915699-06728b80-8329-11eb-8168-4432681ee968.png)

**Comment by Lex-2008** [comment]
Да, по-моему, у тебя for'ов мало (нет, это не сарказм - я бы сделал 5 вложенных for'ов).

Ну хорошо, давай для начала разделим задачу на две:
1) выделим отдельную функцию is_годный_квадрат(размер, начальный элемент), которая будет проверять один квадрат, и return 1 или 0.
2) будем её правильно вызывать. Если в матрице 3 строчки и 4 столбика, то сколько квадратов 3\*3 в ней надо проверить?

**Comment by uliss4ever** [comment]
>1. выделим отдельную функцию is_годный_квадрат(размер, начальный элемент), которая будет проверять один квадрат, и return 1 или 0.
   2.будем её правильно вызывать. Если в матрице 3 строчки и 4 столбика, то сколько квадратов 3*3 в ней надо проверить?

До п.2 пока не дошло, хотя там проверить надо 2 квадрата (то есть столько: сколько начальных точек, а их в данном случае две: f[0][0] и f[0][1]
Не получается сдвинуться никуда. Я хочу чтобы он проверил все элементы и, если нашёл хоть один 0, то break и дальше не проверял (чтобы потом дать команду сдвинуться и проверять сразу следующий вариант). Но у меня нормально получилось только если без break. Хотя это логично, команды куда свигаться в случае выхода из текущего цикла, я же не написала.

![Снимок экрана 2021-03-12 в 13 26 53](https://user-images.githubusercontent.com/32711430/110927582-e053e800-8336-11eb-977e-0edc588737c7.png)
![Снимок экрана 2021-03-12 в 13 27 20](https://user-images.githubusercontent.com/32711430/110927676-fe214d00-8336-11eb-9723-b268c8242c68.png)

**Comment by Lex-2008** [comment]
Хмм, да, давай лучше с другой стороны зайти попробуем.

>> Если в матрице 3 строчки и 4 столбика, то сколько квадратов 3\*3 в ней надо проверить?
>
>проверить надо 2 квадрата (то есть столько: сколько начальных точек, а их в данном случае две: f[0][0] и f[0][1]

Правильно! Значит, если у нас матрица 3х4 (3 строчки и 4 столбика), то в ней помещается только 2 квадрата 3x3. **2** квадрата _по горизонтали_, в **1** ряд _по вертикали_.

А что насчет квадратов 2x2 и 1x1? В матрице 3х4, сколько квадратов 2x2 и 1x1 помещается по горизонтали и по вертикали?

Сможешь написать формулу?

**Comment by uliss4ever** [comment]
Да, попробую, но вот на первом скрине. Выяснила сейчас, что считает неправильно. Я бы хотела сделать, чтобы если в квадрате  все единицы, возвращало 1, а если есть хоть один 0 - то None.

**Comment by Lex-2008** [comment]
Да, у тебя return внутрь for'а залез ;)

**Comment by uliss4ever** [comment]
>Да, у тебя return внутрь for'а залез ;)

О, пасиб, получилось. Теперь есть функция "годный квадрат", но она видимо самым последним циклом будет?
>А что насчет квадратов 2x2 и 1x1? В матрице 3х4, сколько квадратов 2x2 и 1x1 помещается по горизонтали и по вертикали?

квадратов 2 * 2 и 1 * 1 будет (w-1) * (h-1) и w * h соответственно. Где w и h - это ширина и высота матрицы, так?

**Comment by uliss4ever** [comment]
Сейчас думаю, как эту формулу сделать универсальной

**Comment by Lex-2008** [comment]
Что насчет квадратов 3x3, 4х4, 5х5?... Если размер квадрата обозначить буквой n?

**Comment by uliss4ever** [comment]
Пока n <= min (h, w), count = (h - n + 1) * (w - n + 1) Кажется. На матрице 5 * 5 проверила)

**Comment by Lex-2008** [comment]
Да, для квадрата 3х3 и матрицы 3х4 тоже правильно :)

**Comment by Lex-2008** [comment]
Итак, продолжим:

> Теперь есть функция "годный квадрат", но она видимо самым последним циклом будет?

Неет, мы теперь будем вызывать её для каждого из этих квадратов:

> квадратов 2 \* 2 и 1 \* 1 будет (w-1) \* (h-1) и w * h соответственно. Где w и h - это ширина и высота матрицы, так?

Точно! А чем эти квадраты будут отличаться друг от друга? Ну, в смысле, расположением в матрице. Значит, надо функцию "годный квадрат" вызвать для каждого из них - так, чтобы она рассмотрела каждый квадрат по очереди. И так, чтобы она смогла отличить один квадрат от другого. Итак, какие параметры надо в неё передать?

Она у нас пока не очень правильно написана. Ей надо больше параметров передавать.

**Comment by uliss4ever** [comment]
Теперь n надо запихнуть в какой-то for, чтобы оно было всеми значениями, которые только можно.
>Она у нас пока не очень правильно написана. Ей надо больше параметров передавать.

Изо всех сил думаю. Знаешь, кто-то сказал, что люди жалуются на свою память, и хоть бы кто-то - на ум. Вот, пожалуйста, я могу, очень даже могу))

**Comment by uliss4ever** [comment]
>Итак, какие параметры надо в неё передать?

Мне кажется n, h и w. А не только минимум. Сейчас попробую переделать функцию)

**Comment by Lex-2008** [comment]
да, и что такое h и w?

**Comment by uliss4ever** [comment]
>что такое h и w?

Это высота и ширина матрицы.
Нет, теперь другая версия. Нам надо вызвать функцию "годный квадрат" из всех возможных точек в матрице. То есть каждая точка в ней может быть стартовой, хотя бы для единственного квадрата 1 * 1, если это правый нижний угол.

**Comment by Lex-2008** [comment]
новая версия? Ну, можно и так. :) Какие аргументы тогда надо передать в функцию "годный квадрат"?

**Comment by uliss4ever** [comment]
Запуталась, я, Лёш( 
Но сейчас попробую с самого начала. 
>Какие аргументы тогда надо передать в функцию "годный квадрат"?

Полагаю так. Мы не сможем проверить квадрат без n - размер стороны квадрата. Который собственно моеж начинаться от 1 до длины минимальной сторона из этих двух величин - h и w - ширины и высоты матрицы. И стартовая точка нужна.  Но мы же её и так задаём, не в аргументах, а в цикле.

**Comment by Lex-2008** [comment]
> Запуталась, я, Лёш(

Давай тогда вернемся к предыдущей версии, если там было проще :)

> И стартовая точка нужна. Но мы же её и так задаём, не в аргументах, а в цикле.

Да. Я просто предлагаю саму "проверку квадрата" вынести в отдельную функцию. Давай сначала решим задачу попроще: просто перечислим все квадраты. (а потом для каждого квадрата вызовем функцию "проверить квадрат")

Но перед этим решим задачу ещё проще: посчитаем их. Вон, у тебя уже есть формула для count'а!

**Comment by uliss4ever** [comment]
Там да, проще)
1. count = (h - n + 1) * (w - n + 1)
2. Вот тут я и ломаюсь. Как добраться до всех стартовых точек? for i in range (len(matrix)): - это попасть ко всем рядам матрицы. А к элементам каждого ряда?
Просто не понимаю, как это писать. Это вторая двумерная задача, первая была с конём и она была гораздо проще.

Хотя к элементам я тоже могу попасть: for j in range(len(i)):

**Comment by uliss4ever** [comment]
for k in range(1, n+1) - это мы перечислим все возможные стороны квадрата.

**Comment by uliss4ever** [comment]
Да вот добавила, но тут такое не прокатывает) Просто я всегда когда решаю, этот класс Solution и селф убираю, а потом на сайте добавляю обратно. Если решение в одну функцию, то прокатывает такое. Вообще это не так уже важно, что на сайте. Считает оно в пичарме точно корректно. Можно конечно, забить. С другой стороны - это то же самое ООП, я его что тут не понимаю, что в других местах (хотя уже гораздо, намного больше понимаю, благодаря тебе). Вот  такая красота. Кстати, из хорошего - я убрала вызов функции "годный квадрат" и оставила только проверку. И ничего не сломалось!
![Снимок экрана 2021-03-12 в 22 31 04](https://user-images.githubusercontent.com/32711430/110989609-36e71380-8383-11eb-950f-fb93e43465ee.png)

**Comment by uliss4ever** [comment]
Вот, только в 26 строчке осталось.
![Снимок экрана 2021-03-12 в 22 35 44](https://user-images.githubusercontent.com/32711430/110989752-5ed67700-8383-11eb-9533-9b60cd221a54.png)

**Comment by Lex-2008** [comment]
О, да, правильно! Тот годный квадрат ничего не делал :)
